src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+
+            for (auto i = childCount - 2; i > 0; --i) {
+                tempAst = AnalyserEquationAst::create();
+
+                tempAst->mPimpl->populate(AnalyserEquationAst::Type::PIECEWISE, astParent);
+
+                processNode(mathmlChildNode(node, i), tempAst->mPimpl->mLeftChild, tempAst, component, equation);
+
+                astRight->mPimpl->mParent = tempAst;
+
+                tempAst->mPimpl->mRightChild = astRight;
+                astRight = tempAst;
+            }
+
+            astRight->mPimpl->mParent = ast;
+
+            ast->mPimpl->mRightChild = astRight;
+        }
+    } else if (node->isMathmlElement("piece")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::PIECE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+        processNode(mathmlChildNode(node, 1), ast->mPimpl->mRightChild, ast, component, equation);
+    } else if (node->isMathmlElement("otherwise")) {
+        ast->mPimpl->populate(AnalyserEquationAst::Type::OTHERWISE, astParent);
+
+        processNode(mathmlChildNode(node, 0), ast->mPimpl->mLeftChild, ast, component, equation);
+
+        // Token elements.
+
+    } else if (node->isMathmlElement("ci")) {
+        auto variableName = node->firstChild()->convertToString();
+        auto variable = component->variable(variableName);
+        // Note: we always have a variable. Indeed, if we were not to have one,
+        //       it would mean that `variableName` is the name of a variable
+        //       that is referenced in an equation, but not defined anywhere,
+        //       something that is not allowed in CellML and will therefore be
+        //       reported when we validate the model.
+
+        // Have our equation track the (ODE) variable (by ODE variable, we mean
+        // a variable that is used in a "diff" element).
+
+        if (node->parent()->firstChild()->isMathmlElement("diff")) {
+            equation->addOdeVariable(analyserVariable(variable));
+        } else if (!(node->parent()->isMathmlElement("bvar")
+                     && node->parent()->parent()->firstChild()->isMathmlElement("diff"))) {
+            equation->addVariable(analyserVariable(variable));
+        }
+
+        // Add the variable to our AST.
+
+        ast->mPimpl->populate(AnalyserEquationAst::Type::CI, variable, astParent);
+    } else if (node->isMathmlElement("cn")) {
+        if (mathmlChildCount(node) == 1) {
+            // We are dealing with an e-notation based CN value.
+
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString() + "e" + node->firstChild()->next()->next()->convertToString(), astParent);
+        } else {
+            ast->mPimpl->populate(AnalyserEquationAst::Type::CN, node->firstChild()->convertToString(), astParent);
+        }
+
+        // Qualifier elements.
+
+    } else if (node->isMathmlElement("degree")) {
```

Comment:
IT is possible to create a MathML statement that has `<degree>` as a child of something other than `<root>` or `<diff>` ? Do we check that anywhere?

Reply:
`<degree>` can be the child of an `<apply>` for a root, and of a `<bvar>` for a diff. At least, for what we support in MathML. `<degree>` can be the child of `<moment>`, for instance, but it's not something that we support.

IOW, I am relying on the MathML DTD to validate a MathML string. From there, I know what we support (as part of libCellML), so it's all good... until proven wrong. :)

Reply:
That's not what the spec says? "* degree MUST be a child of root or diff"

Reply:
Well, the spec is wrong... and I should have spotted it. Argh! :(

Reply:
This is going to be addressed in issue #688.